log: improve error reporting

tbg opened the pull request:
In particular, this should allow errors like "write: no space left on device" to
expose themselves on sentry.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19080)
<!-- Reviewable:end -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/util/log/crash_reporting.go, line 202 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19080#-KwRDnl7KuKVQtC3JgDL:-KwRDnl7KuKVQtC3JgDM:b-pmigbn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4152a99fd839e2fc44f0f4e84dd983751a7cafd/pkg/util/log/crash_reporting.go#L202)):*
> ```Go
> }
> 
> func censor(r interface{}) string {
> ```

s/censor/redact/g

---

*[pkg/util/log/crash_reporting.go, line 203 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19080#-KwRDwyqDajGg1IKZEno:-KwRDwyrTkO4Ekyw-lyt:b-sk02xk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4152a99fd839e2fc44f0f4e84dd983751a7cafd/pkg/util/log/crash_reporting.go#L203)):*
> ```Go
> 
> func censor(r interface{}) string {
> 	handlePathError := func(v *os.PathError) string {
> ```

These local functions look more confusing than just doing these inline in the type switch (as long as they're not reused, and even then you can often handle it with a fallthrough (handleSafeType looks like the only exception)). 

---

*[pkg/util/log/crash_reporting_test.go, line 73 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19080#-KwRGC_5zJ9mPw0eICpu:-KwRGC_5zJ9mPw0eICpv:b-tveccl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4152a99fd839e2fc44f0f4e84dd983751a7cafd/pkg/util/log/crash_reporting_test.go#L73)):*
> ```Go
> 		},
> 		{
> 			format: "", rs: []interface{}{os.NewSyscallError("write", syscall.Errno(28))},
> ```

Errno values are not portable; use `errno.ENOSPC`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19080#-:-KwRGWdpZNpc3l6M6Uzd:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
log: improve error reporting

In particular, this should allow errors like "write: no space left on device" to
expose themselves on sentry.

tschottdorf commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/util/log/crash_reporting.go, line 202 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19080#-KwRDnl7KuKVQtC3JgDL:-Kw_NMu-plLHI6cEznp3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4152a99fd839e2fc44f0f4e84dd983751a7cafd/pkg/util/log/crash_reporting.go#L202)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/censor/redact/g
</blockquote></details>

Done.

---

*[pkg/util/log/crash_reporting.go, line 203 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19080#-KwRDwyqDajGg1IKZEno:-Kw_NO1aUDA5q7ZxgSvv:bvjbepa) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4152a99fd839e2fc44f0f4e84dd983751a7cafd/pkg/util/log/crash_reporting.go#L203)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

These local functions look more confusing than just doing these inline in the type switch (as long as they're not reused, and even then you can often handle it with a fallthrough (handleSafeType looks like the only exception)). 
</blockquote></details>

Done. I opted to avoid fallthroughs though.

---

*[pkg/util/log/crash_reporting_test.go, line 73 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19080#-KwRGC_5zJ9mPw0eICpu:-Kw_NnngTC5suhDiMmS4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4152a99fd839e2fc44f0f4e84dd983751a7cafd/pkg/util/log/crash_reporting_test.go#L73)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Errno values are not portable; use `errno.ENOSPC`. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19080)*
<!-- Sent from Reviewable.io -->
